storagemarket/impl/clientstates/client_states.go

```diff
@@ -0,0 +1,122 @@
+package clientstates
+
+import (
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	"github.com/filecoin-project/go-fil-markets/storagemarket/impl/clientutils"
+	"github.com/filecoin-project/go-fil-markets/storagemarket/network"
+	smnet "github.com/filecoin-project/go-fil-markets/storagemarket/network"
+	"github.com/filecoin-project/go-statemachine/fsm"
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+)
+
+var log = logging.Logger("storagemarket_impl")
+
+// ClientDealEnvironment is an abstraction for interacting with
+// dependencies from the storage client environment
+type ClientDealEnvironment interface {
+	Node() storagemarket.StorageClientNode
+	DealStream(proposalCid cid.Cid) (smnet.StorageDealStream, error)
+	CloseStream(proposalCid cid.Cid) error
+}
+
+// ClientStateEntryFunc is the type for all state entry functions on a storage client
+type ClientStateEntryFunc func(ctx fsm.Context, environment ClientDealEnvironment, deal storagemarket.ClientDeal) error
+
+// EnsureFunds attempts to ensure the client has enough funds for the deal being proposed
+func EnsureFunds(ctx fsm.Context, environment ClientDealEnvironment, deal storagemarket.ClientDeal) error {
+	if err := environment.Node().EnsureFunds(
+		ctx.Context(), deal.Proposal.Client, deal.Proposal.Client, deal.Proposal.ClientBalanceRequirement()); err != nil {
+		return ctx.Trigger(storagemarket.ClientEventEnsureFundsFailed, err)
+	}
+
+	return ctx.Trigger(storagemarket.ClientEventFundsEnsured)
+}
+
+// ProposeDeal sends the deal proposal to the provider
+func ProposeDeal(ctx fsm.Context, environment ClientDealEnvironment, deal storagemarket.ClientDeal) error {
+	s, err := environment.DealStream(deal.ProposalCid)
+	if err != nil {
+		return ctx.Trigger(storagemarket.ClientEventDealStreamLookupErrored, err)
+	}
+
+	proposal := network.Proposal{DealProposal: &deal.ClientDealProposal, Piece: deal.DataRef}
+	if err := s.WriteDealProposal(proposal); err != nil {
+		return ctx.Trigger(storagemarket.ClientEventWriteProposalFailed, err)
+	}
+
+	return ctx.Trigger(storagemarket.ClientEventDealProposed)
+}
+
+// VerifyDealResponse reads and verifies the response from the provider to the proposed deal
```

Comment:
Should read, "VerifyDealResponse waits, waits, and waits, and waits some more, and then reads and verifies..."  :)

Reply:
omg... I am very ready to send that intent to accept a lot earlier and switch up the data transfer initiation... just gotta finish this here refactor first.